nemoguardrails/library/ai_defense/actions.py

```
@@ -0,0 +1,116 @@
+    is_blocked = result.get("is_blocked", True)
+    return is_blocked
+
+
+@action(is_system_action=True, output_mapping=ai_defense_text_mapping)
+async def ai_defense_inspect(
+    user_prompt: Optional[str] = None, bot_response: Optional[str] = None, **kwargs
+):
+    api_key = os.environ.get("AI_DEFENSE_API_KEY")
+    if api_key is None:
+        msg = "AI_DEFENSE_API_KEY environment variable not set."
+        log.error(msg)
+        raise ValueError(msg)
+
+    api_endpoint = os.environ.get("AI_DEFENSE_API_ENDPOINT")
+    if api_endpoint is None:
+        msg = "AI_DEFENSE_API_ENDPOINT environment variable not set."
+        log.error(msg)
+        raise ValueError(msg)
+
+    headers = {
+        "X-Cisco-AI-Defense-API-Key": api_key,
+        "Content-Type": "application/json",
+        "Accept": "application/json",
+    }
+
+    if bot_response is not None:
+        role = "assistant"
+        text = str(bot_response)
+    elif user_prompt is not None:
+        role = "user"
+        text = str(user_prompt)
+    else:
+        msg = "Either user_prompt or bot_response must be provided"
+        log.error(msg)
+        raise ValueError(msg)
+
+    messages = [{"role": role, "content": text}]
+
+    metadata = None
+    user = kwargs.get("user")
+    if user is not None:
+        metadata = {"user": user}
+
+    payload: Dict[str, Any] = {"messages": messages}
+    if metadata:
+        payload["metadata"] = metadata
+
+    async with httpx.AsyncClient() as client:
+        try:
+            resp = await client.post(api_endpoint, headers=headers, json=payload)
+            resp.raise_for_status()
+            data = resp.json()
+        except httpx.HTTPStatusError as e:
+            msg = f"Error calling AI Defense API: {e}"
+            log.error(msg)
+            raise ValueError(msg)
+
+        # Compose a consistent return structure for flows
+        is_safe = bool(data.get("is_safe", True))
```

Any concerns with a malformed response dict being treated as safe by default?

---

Added configuration for fail_open vs fail closed. And also a timeout config.